File: 1004-max-consecutive-ones-iii/1004-max-consecutive-ones-iii.java
class Solution {
    public int longestOnes(int[] nums, int k) {
     
        int l=1;
        int h=nums.length;
        
        int ans=0;
        
        while(l<=h){
         
            int mid=(h+l)/2;
            
            boolean ans2=longestOnes(nums,k,mid);
            
            if(ans2){
                ans=mid;
                l=mid+1;
            }else{
                h=mid-1;
            }            
        }
        
        return ans;
    }
    
    public static boolean longestOnes(int[] nums,int k,int mid){
        
        int one=0;
        int n=nums.length;
        for(int i=0;i<mid;i++){
            one+=nums[i];
        }
        
        if(one+k>=mid){
         return true;   
        }
        
        for(int i=1;i<=n-mid;i++){
        
            one-=nums[i-1];
            one+=nums[i+mid-1];
        
        if(one+k>=mid){
         return true;   
        }   
        }
        
        return false;
        
    }
}

File: 1004-max-consecutive-ones-iii/1004-max-consecutive-ones-iii.cpp
class Solution {
    private:
    bool possible(vector<int> &nums,int mid,int k)
    {
        int n =nums.size();
      int count[2]={0};
      for(int i=0;i<mid;i++){
          count[nums[i]]++;
      }  
        
      if(count[0]<=k)return true;
      for(int i=mid;i<n;i++)
      {
          count[nums[i]]++;
          
          count[nums[i-mid]]--;
          
          if(count[0]<=k)return true;
      }
       return false;
    }
        
    
public:
    int longestOnes(vector<int>& nums, int k) {
      int n= nums.size();
      int ans  = n;
      int l=0,r=n;
      while(l<=r)
      {
          int mid = l+(r-l)/2;
          if(possible(nums,mid,k)==true){
              ans=mid;
              l=mid+1;
            }
          else{
            r=mid-1;  
          }
            
      }
        return ans;
    }
};